src/Geometry/K3Auto.jl

```diff
@@ -0,0 +1,2059 @@
+  if  r == 26
+    @req lw == 0 "not a weyl vector"
+  end
+  if  r == 18
+    @req lw == 620 "not a weyl vector"
+  end
+  if  r == 10
+    @req lw == 1240 "not a weyl vector"
+  end
+  # transform L to have the standard basis
+  # we assume that the basis of L is obtained by completing a basis of R
+  # hence we can throw away the R coordinates of a weyl vector when projecting to S
+  R = lll(Hecke.orthogonal_submodule(L, S))
+
+  # the following completes the basis of R to a basis of L
+  basisRL = solve_left(basis_matrix(L),basis_matrix(R))
+  basisRL = change_base_ring(ZZ, basisRL)
+
+  A, j = snf(abelian_group(basisRL))
+  T = reduce(vcat, [j(i).coeff for i in gens(A)])
+  basisL1 = vcat(T, basisRL)*basis_matrix(L)
+
+  # carry the weyl vector along
+  L1 = lattice(ambient_space(L), basisL1)
+  weyl = change_base_ring(ZZ, solve_left(basisL1, weyl*basis_matrix(L)))
+  basisSL1 = solve_left(basis_matrix(L1), basis_matrix(S))
+  basisRL1 = solve_left(basis_matrix(L1), basis_matrix(R))
+
+  # Assure that L has the standard basis.
+  L = Zlattice(gram=gram_matrix(L1))
+  V = ambient_space(L)
+  # carry S along
+  S = lattice(V, basisSL1)
+  R = lattice(V, basisRL1)
+  @req is_S_nondegenerate(L, S, change_base_ring(QQ,weyl)) "Weyl vector is S degenerate"
+
+  SS = Zlattice(gram=gram_matrix(S))
+
+  # precomputations
+
+  @assert iszero(basis_matrix(R)[1:end,1:rank(S)])
+  bSR = vcat(basis_matrix(S),basis_matrix(R))
+  ibSR = inv(bSR)
+  I = identity_matrix(QQ,degree(L))
+  # prS: L --> S^\vee given with respect to the standard basis of L and the basis of S
+  prS = ibSR*I[:,1:rank(S)]#*basis_matrix(S)
+  @assert prS[rank(S)+1,:]==0
+
+
+  if compute_OR
+    dd = diagonal(gram_matrix(R))
+    @vprint :K3Auto 2 "computing orthogonal group\n"
+    OR = orthogonal_group(R)
+    @vprint :K3Auto 2 "done\n"
+    DR = discriminant_group(R)
+    ODR = orthogonal_group(DR)
+    imOR = [ODR(hom(DR,DR,[DR(lift(d)*f) for d in gens(DR)])) for f in gens(OR)]
+    DS = discriminant_group(S)
+    DSS = discriminant_group(SS)
+    ODSS = orthogonal_group(DSS)
+    orderimOR = order(sub(ODR,imOR)[1])
+    @vprint :K3Auto 1 "[O(S):G] = $(order(ODSS)//orderimOR)\n"
+    if order(ODR)== orderimOR
+      membership_test = (g->true)
+    else
+      phiSS_S = hom(DSS,DS,[DS(lift(x)*basis_matrix(S)) for x in gens(DSS)])
+      phi,i,j = glue_map(L,S,R)
+      phi = phiSS_S*inv(i)*phi*j
+      img,_ = sub(ODSS,[ODSS(phi*hom(g)*inv(phi)) for g in imOR])
+      ds = degree(SS)
+      membership_test = (g->ODSS(hom(DSS,DSS,[DSS(vec(matrix(QQ, 1, ds, lift(x))*g)) for x in gens(DSS)])) in img)
+    end
+  else
+    membership_test(g) = is_pm1_on_discr(SS,g)
+  end
+
+  d = exponent(discriminant_group(S))
+  Rdual = dual(R)
+  sv = short_vectors(rescale(Rdual,-1), 2)
+  # not storing the following for efficiency
+  # append!(sv,[(-v[1],v[2]) for v in sv])
+  # but for convenience we include zero
+  T = typeof(sv).parameters[1].parameters[1].parameters[1]
```

Comment:
Huh, this looks nasty (no offense). What does it do? Perhaps a comment could be added to explain that? It is not immediately obvious to me what type `sv` has and hence what this does.

Reply:
Now that the interface for `short_vectors` is finalized I can rewrite this part. Thank you for pointing me to it.